Parcel v2: SVGO + svg to react transformers

green-arrow opened the pull request:
<!---
Thanks for filing a pull request 😄 ! Before you submit, please read the following:

Search open/closed issues before submitting since someone might have pushed the same thing before!
-->

# ↪️ Pull Request

In Parcel v1, I used `parcel-plugin-svgr` to generate react components out of `.svg` files. I thought it might be nice to provide a native transformer for this in Parcel v2.

Related issue: #4105 

## 💻 Examples

Add the relevant transforms to your `.parcelrc`:

```json
{
  "transformers": {
    "*.svg": ["@parcel/transformer-svgo", "@parcel/transformer-svg-react", "@parcel/transformer-js"]
  }
}
```

## 🚨 Test instructions

In a react application, add the transformer to the `.parcelrc`, import an SVG file, and render it as a component:

```jsx
import MyLogo from './Logo.svg';

export default function() {
  return (<header>
    <MyLogo title="Any react SVG props can be passed" />
  </header>);
}
```

## ✔️ PR Todo

- [x] Added/updated unit tests for this change
- [X] Filled out test instructions (In case there aren't any unit tests)
- [X] Included links to related issues/PRs

<!--
Love parcel? Please consider supporting our collective:
👉  https://opencollective.com/parcel/donate
-->

green-arrow commented:
I'm more than happy to write any tests needed if someone can point me in the right direction. My thought was to just put a test under `core/integration/svg-react` and make sure a basic SVG-to-react use case works.

Also, I'm not able to completely test the output of the final bundle on my app because I get this error at the end:

```
 @parcel/runtime-js: bundleGraph.getBundleGroupsReferencedByBundle(...) is not a function or its retur
n value is not iterable
TypeError: bundleGraph.getBundleGroupsReferencedByBundle(...) is not a function or its return value is n
ot iterable
    at Object.apply (/Users/andrewwa/Developer/workspace/nova/node_modules/@parcel/runtime-js/lib/JSRunt
ime.js:76:22)
    at applyRuntimes (/Users/andrewwa/Code/parcel/packages/core/core/lib/applyRuntimes.js:46:44)
```

Any help with that would also be super appreciated 😄

green-arrow commented on packages/transformers/svg-react/src/SVGRTransformer.js:
```diff
@@ -0,0 +1,77 @@
+
+const [parcelPluginSymbol] = Object.getOwnPropertySymbols(JSTransformer);
+
+JSTransformer[parcelPluginSymbol].getConfig = async function({asset}) {
```
@DeMoorJasper - My goal was to re-use the logic in `JSTransformer`, since ideally the only difference is this transformer does a "pre-transform" to get it in a react component state. Other than that, it should go through the normal JS transformation. The reason I wanted to do this was to preserve the babel transforms, since I know differential bundling is a key feature of parcel 2.

Is what I said above accurate? If so, is the right approach to somehow extend `JSTransformer`? If not, should I just duplicate the necessary logic from `JSTransformer` in this one?

DeMoorJasper replied:
In Parcel 2 we have a concept called pipelines so in theory there would be no need to extend any transformer as you can let another transformer run before or after this transformer

green-arrow replied:
@DeMoorJasper - I updated the transformer to just generate the react code and not run any JS transforms. In my application, I can then update my `.parcelrc` to run it through both transformers:

```json
{
  "extends": ["@parcel/config-default"],
  "transforms": {
    "*.svg": ["@parcel/transformer-svgo", "@parcel/transformer-svg-react", "@parcel/transformer-js"]
  }
}
```

Is there any way to specify a "follow-up" transformer? For SVG to react, it seems to make sense to always run the result of this transformer through the one that handles JS, so it is transformed for the target environment properly. Thoughts?

green-arrow commented on packages/transformers/svg-react/src/SVGRTransformer.js:
```diff
@@ -0,0 +1,77 @@
+  svgoConfig = svgoConfig || {};
+  svgoConfig = {...defaultConfig, ...svgoConfig};
+
+  this.svgo = new SVGO(svgoConfig);
```
@DeMoorJasper - that's a great point. I'll pull this out into it's own `@parcel/transformer-svgo` if that works

green-arrow commented:
@DeMoorJasper - Recent commit addresses your PR comments.

I'll look into writing the integration test today.

It looks like the failing test is also failing on `v2`, so doesn't appear related to these changes. I can definitely look into it though! Would you want a separate PR for a fix for the test or should I include it here?

DeMoorJasper commented on packages/transformers/svg-react/src/SvgReactTransformer.js:
```diff
@@ -0,0 +1,58 @@
+
+export default new Transformer({
+  async parse({asset}) {
+    let code = await asset.getCode();
```
If you move the logic from parse to transform you don't have to return an ast and can just use asset.setcode and return the asset wihtout needing babel at all.

That's what we've done in a couple transformers that dont' utilise ast's

DeMoorJasper commented on packages/transformers/svgo/src/SVGOTransformer.js:
```diff
@@ -0,0 +1,45 @@
+  },
+
+  async generate({asset}) {
+    let code = await asset.getCode();
```
Maybe move this to transform and not have a generate?

green-arrow replied:
For some reason I thought `generate` was required 😛

DeMoorJasper commented on packages/transformers/svgo/src/SVGOTransformer.js:
```diff
@@ -0,0 +1,45 @@
+    svgoConfig = svgoConfig || {};
+    svgoConfig = {...defaultConfig, ...svgoConfig};
+
+    this.svgo = new SVGO(svgoConfig);
```
Not entirely sure if this is a good idea as it's not guaranteed that this state will be kept in other functions. It currently works but we could shift around a bit in transformrunner and break this.

Probably best to initialise svgo where you actually use it and keep the config to the purest JSON form possible. This will also help us improve this plugin for better caching in the future :)

DeMoorJasper commented:
@green-arrow v2 appears to be stable not sure why it is failing in this PR and not in v2.

green-arrow commented:
@DeMoorJasper - Addressed your feedback in the last commit

Maybe my fork `v2` branch is not up to date (it failed when I ran that integration test locally). I'll give it another go.

green-arrow commented:
@DeMoorJasper - I synced my fork with the parcel `v2` branch, and still receive the same error for that integration test when I run it locally.

DeMoorJasper commented:
Looks great. I'll have a look at the test tomorrow, strange that it fails in here and nowhere else.

green-arrow commented:
@DeMoorJasper - As far as I can tell, something weird is happening with that `matchesEngine` call and I'm not sure exactly what. Something that seems weird, though, is that the `index.html` file used for the test contains the script entry as follows:

```html
<script type="module" src="./index.js"></script>
```

In this case, it seems like the target should be browsers that support modules, which would mean no polyfill necessary. So it seems like the output might be correct, but the test is just using the wrong input file?

I guess the question is, given this type of entry point (HTML with type="module" in the script), along with a target browser query of an older browser (like IE 11), what should happen? I'm not sure I know what the correct answer to that question is, but I think it's the root cause of why this particular test is broken.

DeMoorJasper commented:
@green-arrow seems rebuilding the yarn lockfile and using the current version of parcel in transforms solved the issue, not entirely sure why but it works so no need to dig deeper into it for now.

green-arrow commented:
@DeMoorJasper - Latest commit contains an integration test for `svgo` and `svg-react` transforms. Let me know if you'd like me to add anything else!

DeMoorJasper commented on packages/core/integration-tests/test/svg.js:
```diff
@@ -0,0 +1,39 @@
+    '*.svg': [
+      '@parcel/transformer-svgo',
+      '@parcel/transformer-svg-react',
+      '@parcel/transformer-js',
```
replace this with `...` than it will run the Javascript pipeline instead :)

DeMoorJasper replied:
I just did it myself, hopefully everything works. So we can ship this 🚀

green-arrow replied:
@DeMoorJasper - haha whoops I did it too and didn't realize.

Just curious, what is that magic of `...`?

DeMoorJasper replied:
It's based on the spread operator, it tells parcel to look for any additional pipelines if there is any. and place them at the position of the `...` in this case after svgo and svg-react.

I think this is actually no necessary and will automatically run through the js pipeline without the `...`.

See the PR https://github.com/parcel-bundler/parcel/pull/2568 for more info

mischnic replied:
`...` means "continue with the pipeline declared in the `extend`ed config".

In this case, doing `asset.type = "js";` will switch to the `js` pipeline either way, so `...` would be ignored anyway

green-arrow commented:
Hi @DeMoorJasper! I was just wondering when the next alpha release will be that contains this change. When I have it linked and running locally my build passes, but fails at the bundling stage with the same error from [this comment](https://github.com/parcel-bundler/parcel/issues/4104#issuecomment-584340054). However, the fix there of deleting the cache and re-running does not work for me. I can consistently `rm -rf .parcel-cache && yarn start` and reproduce the error.

I was hoping to get an alpha build to rule out anything weird from my local linked install. Thoughts?

DeMoorJasper commented:
@green-arrow you can use the nightly builds of parcel `yarn add parcel@nightly`

Not entirely sure when we'll release a new alpha, we're probably gonna try and get some caching stability and other things before moving to a new alpha (or beta)

CrashyBang commented:
Out of curiosity are all 3 `["@parcel/transformer-svgo", "@parcel/transformer-svg-react", "@parcel/transformer-js"]` packages required?

Cheers.

mischnic commented:
- svgo: compress (optional)
- svg-react: transform svg into JS exporting a react component (required)
- js: collect dependencies (required)

Your config should look like this:
```json
{
  extends: ['@parcel/config-default'],
  transforms: {
    '*.svg': ['@parcel/transformer-svgo', '@parcel/transformer-svg-react']
  }
}
```
The js transformer will automatically be called.

CrashyBang commented:
Awsome, so in theory:

```
{
  extends: ['@parcel/config-default'],
  transforms: {
    '*.svg': ['@parcel/transformer-svg-react']
  }
}
```

Would work also?

green-arrow commented:
@CrashyBang yes, the SVGO plugin is completely optional.

fregante commented:
I wasn't able to make the easiest version of this work. Have a look: https://github.com/parcel-bundler/parcel/issues/5206
